File: app/src/main/java/me/zhanghai/android/files/provider/document/DocumentFileAttributes.java
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.document;

import android.net.Uri;
import android.os.Parcel;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import me.zhanghai.android.files.provider.common.ParcelableContentProviderFileAttributes;

public class DocumentFileAttributes extends ParcelableContentProviderFileAttributes {

    private final int mFlags;

    DocumentFileAttributes(long lastModifiedTimeMillis, @Nullable String mimeType, long size,
                           int flags, @NonNull Uri uri) {
        this(new DocumentFileAttributesImpl(lastModifiedTimeMillis, mimeType, size, flags, uri));
    }

    private DocumentFileAttributes(@NonNull DocumentFileAttributesImpl attributes) {
        super(attributes);

        mFlags = attributes.getFlags();
    }

    public int getFlags() {
        return mFlags;
    }


    public static final Creator<DocumentFileAttributes> CREATOR =
            new Creator<DocumentFileAttributes>() {
                @Override
                public DocumentFileAttributes createFromParcel(Parcel source) {
                    return new DocumentFileAttributes(source);
                }
                @Override
                public DocumentFileAttributes[] newArray(int size) {
                    return new DocumentFileAttributes[size];
                }
            };

    protected DocumentFileAttributes(Parcel in) {
        super(in);

        mFlags = in.readInt();
    }


    @Override
    public void writeToParcel(Parcel dest, int flags) {
        super.writeToParcel(dest, flags);

        dest.writeInt(mFlags);
    }
}

File: app/src/main/java/me/zhanghai/android/files/provider/document/DocumentFileAttributes.kt
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.document

import android.net.Uri
import android.os.Parcelable
import java8.nio.file.attribute.FileTime
import kotlinx.parcelize.Parcelize
import kotlinx.parcelize.WriteWith
import me.zhanghai.android.files.provider.common.AbstractContentProviderFileAttributes
import me.zhanghai.android.files.provider.common.FileTimeParceler
import org.threeten.bp.Instant

@Parcelize
internal class DocumentFileAttributes(
    override val lastModifiedTime: @WriteWith<FileTimeParceler> FileTime,
    override val mimeType: String?,
    override val size: Long,
    override val fileKey: Parcelable,
    private val flags: Int
) : AbstractContentProviderFileAttributes() {
    fun flags(): Int = flags

    companion object {
        fun from(
            lastModifiedTimeMillis: Long,
            mimeType: String?,
            size: Long,
            flags: Int,
            uri: Uri
        ): DocumentFileAttributes {
            val lastModifiedTime = FileTime.from(Instant.ofEpochMilli(lastModifiedTimeMillis))
            val fileKey = uri
            return DocumentFileAttributes(lastModifiedTime, mimeType, size, fileKey, flags)
        }
    }
}
